compiler-plugin/src/main/java/io/ballerina/lib/data/csvdata/compiler/CsvDataTypeValidator.java

```diff
@@ -0,0 +1,649 @@
+        Location pos = location.orElseGet(() -> currentLocation);
+        DiagnosticInfo diagnosticInfo = new DiagnosticInfo(diagnosticsCodes.getCode(),
+                diagnosticsCodes.getMessage(), diagnosticsCodes.getSeverity());
+        if (pos == null || (allDiagnosticInfo.containsKey(pos) && allDiagnosticInfo.get(pos).equals(diagnosticInfo))) {
+            return;
+        }
+        allDiagnosticInfo.put(pos, diagnosticInfo);
+        ctx.reportDiagnostic(DiagnosticFactory.createDiagnostic(diagnosticInfo, pos));
+    }
+
+    private void processModuleVariableDeclarationNode(ModuleVariableDeclarationNode moduleVariableDeclarationNode,
+                                                      SyntaxNodeAnalysisContext ctx) {
+        Optional<ExpressionNode> initializer = moduleVariableDeclarationNode.initializer();
+        currentLocation = moduleVariableDeclarationNode.typedBindingPattern().typeDescriptor().location();
+        if (initializer.isEmpty()) {
+            return;
+        }
+        ExpressionNode expressionNode = initializer.get();
+        if (!isParseFunctionOfStringSource(expressionNode)) {
+            return;
+        }
+
+        Optional<Symbol> symbol = semanticModel.symbol(moduleVariableDeclarationNode.typedBindingPattern());
+        symbol.map(s -> (VariableSymbol) s).map(VariableSymbol::typeDescriptor)
+                .ifPresent(s -> {
+                    validateExpectedType(s, currentLocation, ctx);
+                    validateFunctionParameterTypes(expressionNode, s, expressionNode.location(), ctx);
+                });
+    }
+
+    private void processTypeDefinitionNode(TypeDefinitionNode typeDefinitionNode, SyntaxNodeAnalysisContext ctx) {
+        Node typeDescriptor = typeDefinitionNode.typeDescriptor();
+        currentLocation = typeDefinitionNode.typeDescriptor().location();
+        if (typeDescriptor.kind() != SyntaxKind.RECORD_TYPE_DESC) {
+            return;
+        }
+        validateRecordTypeDefinition(typeDefinitionNode, ctx);
+    }
+
+    private void validateRecordTypeDefinition(TypeDefinitionNode typeDefinitionNode, SyntaxNodeAnalysisContext ctx) {
+        Optional<Symbol> symbol = semanticModel.symbol(typeDefinitionNode);
+        if (symbol.isEmpty()) {
+            return;
+        }
+        TypeDefinitionSymbol typeDefinitionSymbol = (TypeDefinitionSymbol) symbol.get();
+        detectDuplicateFields((RecordTypeSymbol) typeDefinitionSymbol.typeDescriptor(), ctx);
+    }
+
+    private void detectDuplicateFields(RecordTypeSymbol recordTypeSymbol, SyntaxNodeAnalysisContext ctx) {
+        Set<String> fieldMembers = new HashSet<>();
+        for (Map.Entry<String, RecordFieldSymbol> entry : recordTypeSymbol.fieldDescriptors().entrySet()) {
+            RecordFieldSymbol fieldSymbol = entry.getValue();
+            String name = getNameFromAnnotation(entry.getKey(), fieldSymbol.annotAttachments());
+            if (fieldMembers.contains(name)) {
```

Comment:
I belive the idiomatic way to do this is to simpley use the return value of add method (all java sets return false if the value was already there)

```suggestion
            if (!fieldMembers.add(name)) {
```